### PR TITLE
GRA-1494: Turn client side implementation

### PR DESCRIPTION
**Pull request by abhishek9686**
## Provide link to Netmaker PR if required
https://github.com/gravitl/netmaker/pull/2241
## Provide testing steps

- [x] Take two machines with asymmetric or double nat
- [x] these machines should automatically talk over turn with no connection issues

## Checklist before requesting a review
- [ ] My changes affect only 10 files or less.
- [x] I have performed a self-review of my code and tested it.
- [x] If it is a new feature, I have added thorough tests, my code is <= 1450 lines.
- [ ] If it is a bugfix, my code is <= 200 lines.
- [x] My functions are <= 80 lines.
- [ ] I have had my code reviewed by a peer.
- [ ] My unit tests pass locally.
- [x] Netclient & Netmaker are awesome.

### CODE REVIEW

**Review by mattkasun — CHANGES_REQUESTED** [review]

File: auth/auth.go [hunk 1]
```diff
@@ -1,5 +1,5 @@
-// Package functions provides netclient logic
-package functions
+// Package auth provides netclient auth logic with server
+package auth
```
[the remark this answers is not in the document]
Comment (abhishek9686) [reply auth/auth.go:2]: due to cyclic dependency

File: nmproxy/config/peer_config.go [hunk 2]
```diff
@@ -276,3 +279,87 @@ func (c *Config) GetPeersIDsAndAddrs(server, peerKey string) (map[string]nm_mode
...
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.ifaceConfig.hostTurnCfg
+}
+
+// Config.GetTurnCfg - gets the turn config
+func (c *Config) GetTurnCfg(server string) (t models.TurnCfg, ok bool) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	t, ok = c.ifaceConfig.hostTurnCfg[server]
```
[the remark this answers is not in the document]
Comment (abhishek9686) [reply nmproxy/config/peer_config.go:319]: when you return map[key] it just returns value, second parameter (bool) is skipped

File: nmproxy/turn/turn_helper.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+			t, ok := config.GetCfg().GetPeerTurnCfg(signal.Server, signal.FromHostPubKey)
+			if ok {
+				if signal.TurnRelayEndpoint == "" {
+					continue
+				}
+				// reset
+				if conn, ok := config.GetCfg().GetPeer(signal.FromHostPubKey); ok {
+					if conn.Config.UsingTurn && t.PeerTurnAddr != signal.TurnRelayEndpoint {
+						logger.Log(0, fmt.Sprintf("-------> Turn Peer Addr Has Been Changed From %s to %s", t.PeerTurnAddr, signal.TurnRelayEndpoint))
```
[the remark this answers is not in the document]
Comment (abhishek9686) [reply nmproxy/turn/turn_helper.go:45]: will remove it, added for debugging

File: nmproxy/turn/turn.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
+package turn
```
[the remark this answers is not in the document]
Comment (abhishek9686) [reply nmproxy/turn/turn.go:1]: turn connections are handled by proxy, it makes sense to keep it within the proxy

**Review by afeiszli — APPROVED** [review]

**Review by mattkasun — APPROVED** [review]

**Review by theguy951357 — APPROVED** [review]
pr passes
critical list passes

**Pull request merged**